feat(dataset): move/rename files between datasets

mohammad-sdsc opened the pull request:
# Description

`renku dataset mv` command to move files between datasets or rename a single file. To choose which files to move we can use `-I` or `-X` to define a pattern similar to what other `renku dataset` commands accept.

Fixes https://github.com/SwissDataScienceCenter/renku-python/issues/1159

emmjab commented:
Thanks @mohammad-sdsc ! It works as I would expect (e.g. similar to the familiar `mv` command, except dataset must exist first, and great error message when it doesn't already)


help text & commit messages look good: 

```
$ ▶ renku dataset mv --help
Usage: renku dataset mv [OPTIONS] SOURCE_DATASET TARGET_DATASET

  Move matching files from a dataset to another.

Options:
  -I, --include TEXT             Include files matching given pattern.
  -X, --exclude TEXT             Exclude files matching given pattern.
  -d, --dst, --destination TEXT  Destination file or directory within the
                                 target dataset

  -y, --yes                      Confirm overwriting of existing files.
  -h, --help                     Show this message and exit.
```
```
commit 09d8f3ff594c131ffefcd9f2a2f1a5cd6b1dd56f
Author: Emma Jablonski <e.jablonski@gmail.com>
Date:   Tue May 5 17:37:45 2020 +0000

    renku dataset: committing 1 moved files

commit 0130723636a94de2102574e8c277aea097a981b2
Author: Emma Jablonski <e.jablonski@gmail.com>
Date:   Tue May 5 17:37:07 2020 +0000

    renku dataset mv penguins new_penguins

commit 259a5e72bce1383f5dab6a859cc8ac1bfdedcbb0
Author: Emma Jablonski <e.jablonski@gmail.com>
Date:   Tue May 5 17:37:07 2020 +0000

    renku dataset: committing 1 moved files

commit d6cd6e053538b6a9df058bd1368bec3e29c69418
Author: Emma Jablonski <e.jablonski@gmail.com>
Date:   Tue May 5 17:36:41 2020 +0000

    renku dataset create new_penguins
```

```
$ ▶ renku dataset mv penguins doesnt_exist_yet
Error: Dataset "doesnt_exist_yet" is not found.
$ ▶ renku dataset mv doesnt_exist_yet penguins
Error: Dataset "doesnt_exist_yet" is not found.
```

And when no files are in the dataset, can't move them, helpful error message: 

```
$ ▶ renku dataset mv new_penguins penguins
Error: Invalid parameter value - No files found.
```

And also the files themselves get juggled between the `data` subdirs -- didn't actually expect this to happen but 👍 

```
$ ▶ ls data/penguins/
emperor.jpg
$ ▶ renku dataset mv penguins new_penguins
$ ▶ ls data/new_penguins/
emperor.jpg
```

If I had added an image to a dataset originally and that file was not in a subdirectory of `data`, if i move it to a new dataset, it gets put into the `data` subdirectory:

```
$ ▶ renku dataset mv penguins -I penguin2.jpg new_penguins
$ ▶ ls data/new_penguins/
emperor.jpg  penguin2.jpg
```

So if i want to put a file that's hanging around somewhere into a `data` subdirectory, can i call mv? Answer:yes.

```
$ ▶ renku dataset mv penguins -I chinstrap.jpg penguins
$ ▶ ls data/penguins/
chinstrap.jpg
```

I think this is all reasonable, but it's a little confusing that `renku dataset add` doesn't put the file into the `data` directory if it exists already in the repo, where `renku dataset mv` does. I think in general the behavior of `renku dataset mv`, moving actual files along with the metadata is more aligned with what new users expect.

mohammad-sdsc commented:
Thanks @emmjab and @Panaetius for the review!

We've decided to copy all data files to a dataset's data directory when adding files. I've created an issue for this: https://github.com/SwissDataScienceCenter/renku-python/issues/1259. The behavior will be the same for `add` and `mv` after this issue is done.

> Maybe since you can also rename files with this command, change to:

I updated the wording for this function and added the behavior when there is no argument (i.e. moving all) to the dataset documentation.

ciyer commented on renku/cli/dataset.py:
```diff
@@ -883,3 +907,42 @@ def update(short_names, creators, include, exclude, ref, delete, external):
     click.secho('OK', fg='green')
+
+
+@dataset.command('mv')
```
I wonder if it isn't better to name the command `mv-files`? This would be more consistent with the commands like `rm`, which operates on datasets (but not files).

I tried to rename a dataset by `mv`, but this didn't work. If the command were `mv-files`, it would be clear that I couldn't rename the dataset using mv.

mohammad-sdsc replied:
We are trying (mostly) to follow git commands names; for example we also have `renku dataset add/unlink` that also work on files. I am not sure if `mv-files` helps here because we normally don't move datasets and can only rename them (renaming is not supported yet BTW).
The main point here is that we use the same `dataset` verb to act on files and datasets. Maybe we should resume the discussion regarding this to separate them (i.e. `renku file` vs `renku dataset`).

ciyer commented on renku/cli/dataset.py:
```diff
@@ -179,6 +179,30 @@
+
+.. code-block:: console
+
+    $ renku dataset mv my-dataset my-dataset -I data.csv -d old-data.csv
```
This is a helpful example because it's a little counterintuitive that to rename a file you have to pass in the same dataset as src and dst. Could you add this to the text that is shown with `renku dataset mv --help`?

mohammad-sdsc replied:
I've fixed this; now, if one doesn't specify a destination dataset then the source dataset is used as destination.

ciyer commented on renku/cli/dataset.py:
```diff
@@ -883,3 +907,42 @@ def update(short_names, creators, include, exclude, ref, delete, external):
+    '-I',
+    '--include',
+    multiple=True,
+    help='Include files matching given pattern.'
```
An example in the help text would be nice. You could show that the pattern needs to be  quoted to prevent shell expansion.

mohammad-sdsc replied:
So far the convention was to keep these help messages short. An example is already in the dataset documentation. We also have an issue to improve UX regarding help in commands: https://github.com/SwissDataScienceCenter/renku-python/issues/1069.

mohammad-sdsc commented:
Update: The last commit allows omitting destination dataset where in that case source is used as destination. This prevents specifying the same dataset name twice when renaming. Also, if source and destination files are the same an error is raised.

mohammad-sdsc commented:
I've changed the wording according to your suggestion.

> Also occurred to me that I didn't test what happens when you try to renku update after moving a dataset that was a participant in a renku run. 

The update command won't work if a file is moved because it uses file's path to find a file. There won't be a trivial fix for that and we should come up with a solution which does not use file paths or one that uses datasets (e.g. #787).

rokroskar commented:
> The update command won't work if a file is moved because it uses file's path to find a file. There won't be a trivial fix for that and we should come up with a solution which does not use file paths or one that uses datasets (e.g. #787).

but then this effectively breaks the project, no? I think we need to find some sort of solution.

mohammad-sdsc commented:
> this effectively breaks the project, no? I think we need to find some sort of solution.

It does if the files are used on some workflows similar to when a file or dataset is removed. To me, a proper solution is to decouple workflows from file paths by assigning fixed IDs to files and refer to them with those IDs independent from their location/name/dataset/etc. #787 would be a good start in this direction.
We can also go through all CWL files and replace old path with new ones (and I guess commit them one-by-one) for the sake of this story. Honestly, I believe this is a patchy solution and is not worth the effort; we can just make it clear in the docs that if users move a file some workflows might break.
Let me know what do you think and/or if you have other solutions in mind.

ciyer commented:
> It does if the files are used on some workflows similar to when a file or dataset is removed. To me, a proper solution is to decouple workflows from file paths by assigning fixed IDs to files and refer to them with those IDs independent from their location/name/dataset/etc. #787 would be a good start in this direction.

I think using git content hashes to identify files is a good idea in the metadata stores, because the metadata apply to the content, the path is not so important, but I think paths make sense in workflows. For example, if you run a script as a workflow and then fix a bug in the script (changing its content hash), you would still want the workflow to run.

> We can also go through all CWL files and replace old path with new ones (and I guess commit them one-by-one) for the sake of this story. Honestly, I believe this is a patchy solution and is not worth the effort; we can just make it clear in the docs that if users move a file some workflows might break.

I agree that doing this automatically for the user is a bad idea. Is this feasible: If a user moves a file, we check if the file is used in any workflows, if so, we ask the user if they want to update the workflows as well?

Panaetius commented on renku/core/management/datasets.py:
```diff
@@ -1209,6 +1211,99 @@ def move_files(self, source_dataset, destination_dataset, files):
+            self._replace_paths_in_workflows(files, commit=commit)
+
+    def _replace_paths_in_workflows(self, files, commit):
+        """Replace file paths in dataflows after moving files."""
```
`s/dataflows/workflows/`?

Panaetius commented on renku/core/management/datasets.py:
```diff
@@ -1209,6 +1211,99 @@ def move_files(self, source_dataset, destination_dataset, files):
+
+            plan = process.association.plan
+
+            for output in plan.outputs or []:
```
Couldn't you use `replace_paths_in_run` here?

Panaetius commented on renku/core/management/datasets.py:
```diff
@@ -1209,6 +1211,99 @@ def move_files(self, source_dataset, destination_dataset, files):
+            replaced = replace_paths_in_process(workflow)
+
+            if replaced:
+                workflow.to_yaml(path=path)
```
you need to add the new paths to the activity index https://github.com/SwissDataScienceCenter/renku-python/blob/master/renku/core/management/repository.py#L486 which is used to find workflows that were involved in a path, and probably remove the old path entries from there as well.

Panaetius commented on renku/core/models/jsonld.py:
```diff
@@ -632,10 +632,11 @@ def asjsonld(self):
         """Store an instance to the referenced YAML file."""
         jsonld_ = self.asjsonld()
-        write_yaml(path=self.__reference__, data=jsonld_)
+        path = path or self.__reference__
```
So far the intended use of this was to wrapp calls to `to_yaml` into `with with_reference():`. I'm not sure if that has any benefit over supplying the path directly (It does some check to make sure the reference wasn't changed). I think we should update to this new version everywhere that is currently wrapped in `with_reference` to make the code more readable.

mohammad-sdsc replied:
I've created an issue for this: https://github.com/SwissDataScienceCenter/renku-python/issues/1361.

mohammad-sdsc commented:
A similar functionality with easier UX is implemented in https://github.com/SwissDataScienceCenter/renku-python/pull/1993. The other PR does not update workflow files; followup story: https://github.com/SwissDataScienceCenter/renku-python/issues/445.
